File: semester_two/algoDat/public/code/vorlesung-12/StringSearch.cpp
// Copyright 2015, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Author: Hannah Bast <bast@cs.uni-freiburg.de>.

#include <vector>
#include <string>
#include "./StringSearch.h"

// _____________________________________________________________________________
std::vector<size_t> StringSearch::findMatchesNaive(const std::string& text,
                                                   const std::string& pattern) {
  numComparisons = 0;
  std::vector<size_t> matches;
  if (pattern.size() == 0 || pattern.size() > text.size()) return matches;
  size_t i = 0;
  while (i < text.size() - pattern.size() + 1) {
    // Check if pattern matches at position i in text.
    size_t j = 0;
    while (j < pattern.size()) {
      numComparisons++;
      if (pattern[j] != text[i + j]) break;
      j++;
    }
    // If loop went to end, pattern matched at position i.
    if (j == pattern.size()) {
      matches.push_back(i);
    }
    i++;
  }
  return matches;
}

// _____________________________________________________________________________
std::vector<size_t> StringSearch::findMatchesKmp(const std::string& text,
                                                 const std::string& pattern) {
  numComparisons = 0;
  std::vector<size_t> matches;
  computeShiftsKmp(pattern);
  if (pattern.size() == 0 || pattern.size() > text.size()) return matches;
  size_t i = 0;
  size_t shift = 0;
  while (i < text.size() - pattern.size() + 1) {
    // Check if pattern matches at position i in text.
    size_t j = shift;
    while (j < pattern.size()) {
      numComparisons++;
      if (pattern[j] != text[i + j]) break;
      j++;
    }
    // If loop went to end, pattern matched at position i.
    if (j == pattern.size()) {
      matches.push_back(i);
    }
    // Here is the only difference to the naive algorithm.
    shift = j > 0 ? shifts[j -1] : 0;
    i = j > 0 ? i + j - shift : i + 1;
    // i++;
  }
  return matches;
}

// _____________________________________________________________________________
void StringSearch::computeShiftsKmp(const std::string& pattern) {
  shifts.resize(pattern.size());
  if (pattern.size() == 0) return;
  shifts[0] = 0;
  for (size_t j = 1; j < pattern.size(); j++) {
    size_t k = shifts[j - 1];
    if (pattern[j] == pattern[k]) {
      shifts[j] = k + 1;
    } else if (pattern[j] == pattern[0]) {
      shifts[j] = 1;
    } else {
      shifts[j] = 0;
    }
  }
}

File: semester_two/algoDat/public/code/vorlesung-12/vorlagen/java/StringSearch.java
/**
 * Copyright 2017, University of Freiburg,
 * Chair of Algorithms and Data Structures.
 * Author: Axel Lehmann <lehmann@cs.uni-freiburg.de>.
 */
import java.util.ArrayList;

/**
 * Collection of string matching functions.
 */
public class StringSearch {

  /**
   * Number of comparisons needed.
   */
  protected int numComparisons;

  /**
   * The shifts for KMP.
   */
  protected int[] shifts;

  /**
   * Construct an empty instance of StringSearch.
   */
  public StringSearch() {
    numComparisons = 0;
    shifts = new int[0];
  }

  /**
   * Find all matches of the given pattern in the given text using the naive
   * algorithm.
   */
  public ArrayList<Integer> findMatchesNaive(String text, String pattern) {
    numComparisons = 0;
    ArrayList<Integer> matches = new ArrayList<>();
    if (pattern.length() == 0 || pattern.length() > text.length()) {
      return matches;
    }

    int i = 0;
    while (i < text.length() - pattern.length() + 1) {
      // Check if pattern matches at position i in text.
      int j = 0;
      while (j < pattern.length()) {
        numComparisons++;
        if (pattern.charAt(j) != text.charAt(i + j)) {
          break;
        }
        j++;
      }
      // If loop went to end, pattern matched at position i.
      if (j == pattern.length()) {
        matches.add(i);
      }
      i++;
    }
    return matches;
  }

  /**
   * Find all matches of the given pattern in the given text using KMP.
   */
  public ArrayList<Integer> findMatchesKmp(String text, String pattern) {
    numComparisons = 0;
    ArrayList<Integer> matches = new ArrayList<>();
    computeShiftsKmp(pattern);
    if (pattern.length() == 0 || pattern.length() > text.length()) {
      return matches;
    }

    int i = 0;
    int shift = 0;
    while (i < text.length() - pattern.length() + 1) {
      // Check if pattern matches at position i in text.
      int j = shift;
      while (j < pattern.length()) {
        numComparisons++;
        if (pattern.charAt(j) != text.charAt(i + j)) {
          break;
        }
        j++;
      }
      // If loop went to end, pattern matched at position i.
      if (j == pattern.length()) {
        matches.add(i);
      }
      // Here is the only difference to the naive algorithm.
      shift = j > 0 ? shifts[j - 1] : 0;
      i = j > 0 ? i + j - shift : i + 1;
    }
    return matches;
  }

  /**
   * Precompute shifts for KMP.
   */
  public void computeShiftsKmp(String pattern) {
    shifts = new int[pattern.length()];
    if (pattern.length() < 1) {
      return;
    }
    shifts[0] = 0;
    for (int j = 1; j < pattern.length(); j++) {
      int k = shifts[j - 1];
      if (pattern.charAt(j) == pattern.charAt(k)) {
        shifts[j] = k + 1;
      } else if (pattern.charAt(j) == pattern.charAt(0)) {
        shifts[j] = 1;
      } else {
        shifts[j] = 0;
      }
    }
  }
}
